src/cr/cube/measures/scale_means.py

```diff
@@ -68,13 +68,24 @@ def margin(self, axis):
         return np.sum(values * margin) / total
 
     def valid_indices(self, axis):
+        # --there is an interaction with CrunchCube._fix_shape() which
```

Comment:
We might want to 🎫 this abstraction. I'm thinking of either `collapsed_dimensions` or `hidden_dimensions` or both (in respective contexts). But yeah, this abstraction definitely needs to happen.

Reply:
The vague idea I have (vague still because I haven't gotten my head completely around the required complexity yet) is that we need some pre-processed working objects that encapsulate a lot of the differences that are now spread all through the code.

One example of this might be `_RawCube` vs. `_SquashedCube`. A `_RawCube` object would be formed directly from the underlying result dict and would know nothing about pruning, MR dimension squashing, or HS-insertion. It would be the source that `_SquashedCube` pulls value from. (Both of these would be internal implementation details, hence the leading underscore.) `_SquashedCube` would be constructed by a factory by passing parameters like `show_empty` and `show_hs` (or whatever that option is), and would then encapsulate providing a consistent "squashed" cube that was appropriately pruned, mr-squashed, etc. (including that its dimensions were squashed in a consistent way) such that clients of `_SquashedCube` (like CrunchCube and CubeSlice) didn't have to be carrying all those "flags" around with them and distributing the prune-vs-show-empty and mr-squashing logic through their code (which shows up as all these if-statements).

Anyway, a topic to discuss further.